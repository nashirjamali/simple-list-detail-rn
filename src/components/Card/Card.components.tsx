import React, {FC} from 'react';
import {TouchableOpacity, View} from 'react-native';

import type {CardProps, CardVariantsType} from '../../@types';
import {
  CARD_VARIANTS,
  GENERAL_TEXT_VARIANTS,
  LABEL_VARIANTS,
} from '../../constants';
import {formatDateToIndonesian, formatToIDR} from '../../utils';

import Label from '../Label/Label.components';
import GeneralText from '../GeneralText/GeneralText.components';

import styles from './Card.styles';
import ArrowRightIcon from '../Icons/ArrowRightIcon.icons';
import DotIcon from '../Icons/DotIcon.icons';

/**
 * Renders a colored strip on the card based on the status.
 *
 * @param {CardVariantsType} status - The status of the card.
 * @returns {JSX.Element} The rendered strip element.
 */
const _renderCardStrip = (status: CardVariantsType): JSX.Element => (
  <View style={styles[status]} />
);

/**
 * Renders a label based on the card status.
 *
 * @param {CardVariantsType} status - The status of the card.
 * @returns {JSX.Element} The rendered label element.
 */
const _renderLabel = (status: CardVariantsType): JSX.Element => {
  switch (status) {
    case CARD_VARIANTS.PENDING:
      return <Label text="Pengecekan" variant={LABEL_VARIANTS.PENDING} />;

    case CARD_VARIANTS.SUCCESS:
      return <Label text="Berhasil" variant={LABEL_VARIANTS.SUCCESS} />;

    default:
      return <Label text="Pengecekan" variant={LABEL_VARIANTS.PENDING} />;
  }
};

/**
 * Renders the bank transfer information.
 *
 * @param {string} senderBank - The bank of the sender.
 * @param {string} beneficiaryBank - The bank of the beneficiary.
 * @returns {JSX.Element} The rendered bank transfer information.
 */
const _renderBankTransfer = (
  senderBank: string,
  beneficiaryBank: string,
): JSX.Element => (
  <View style={styles.bankTransfer}>
    <GeneralText variant={GENERAL_TEXT_VARIANTS.HEADER}>
      {senderBank.toUpperCase()}
    </GeneralText>
    <ArrowRightIcon />
    <GeneralText variant={GENERAL_TEXT_VARIANTS.HEADER}>
      {beneficiaryBank.toUpperCase()}
    </GeneralText>
  </View>
);

/**
 * Renders the main content of the card.
 *
 * @param {CardProps} props - The properties of the card.
 * @returns {JSX.Element} The rendered content of the card.
 */
const _renderContent = ({
  amount,
  beneficiaryBank,
  beneficiaryName,
  completedAt,
  senderBank,
  status,
}: CardProps): JSX.Element => (
  <View style={styles.contentContainer}>
    <View>
      {_renderBankTransfer(senderBank, beneficiaryBank)}
      <GeneralText variant={GENERAL_TEXT_VARIANTS.NORMAL}>
        {beneficiaryName.toUpperCase()}
      </GeneralText>
      <View style={styles.amountAndDate}>
        <GeneralText variant={GENERAL_TEXT_VARIANTS.NORMAL}>
          {formatToIDR(amount)}
        </GeneralText>
        <DotIcon />
        <GeneralText variant={GENERAL_TEXT_VARIANTS.NORMAL}>
          {formatDateToIndonesian(completedAt)}
        </GeneralText>
      </View>
    </View>
    {_renderLabel(status)}
  </View>
);

/**
 * Card component displays transaction information.
 *
 * @param {CardProps} props - The properties of the card.
 * @returns {JSX.Element} The rendered card component.
 */
const Card: FC<CardProps> = (props: CardProps): JSX.Element => (
  <TouchableOpacity style={styles.container}>
    {_renderCardStrip(props.status)}
    {_renderContent(props)}
  </TouchableOpacity>
);

export default Card;
