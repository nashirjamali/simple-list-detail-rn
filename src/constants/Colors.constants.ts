export const COLORS = {
  BRAND: '#FB6544',
  BLACK: '#121212',
  SUCCESS: '#55B585',
  GRAY: '#999999',
  WHITE: '#FFFFFF',
} as const;
