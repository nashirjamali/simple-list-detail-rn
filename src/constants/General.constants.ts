export const API_URL =
  'https://recruitment-test.flip.id/frontend-test' as const;
