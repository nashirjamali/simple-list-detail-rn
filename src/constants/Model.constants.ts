export const TRANSACTION_STATUS = {
  SUCCESS: 'SUCCESS',
  PENDING: 'PENDING',
} as const;
