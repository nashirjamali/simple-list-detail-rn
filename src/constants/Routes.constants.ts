export const ROUTES = {
  TRANSACTION_LIST: 'transactionsList',
  TRANSACTION_DETAILS: 'transactionsDetails',
} as const;
