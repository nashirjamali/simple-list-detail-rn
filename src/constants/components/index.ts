export * from './GeneralText.constants';
export * from './Button.constants';
export * from './Label.constants';
export * from './Card.constants';
