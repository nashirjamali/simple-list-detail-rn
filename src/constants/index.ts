export * from './components';
export * from './Colors.constants';
