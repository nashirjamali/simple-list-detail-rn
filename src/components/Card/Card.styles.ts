import {StyleSheet} from 'react-native';

import type {CardStyles} from '../../@types';

import {COLORS} from '../../constants';

const styles: CardStyles = StyleSheet.create({
  container: {
    borderRadius: 20,
    backgroundColor: COLORS.WHITE,
    display: 'flex',
    flexDirection: 'row',
    width: '100%',
  },
  contentContainer: {
    padding: 20,
    width: '100%',
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  pending: {
    width: 10,
    backgroundColor: COLORS.BRAND,
    borderTopLeftRadius: 20,
    borderBottomLeftRadius: 20,
  },
  success: {
    width: 10,
    backgroundColor: COLORS.SUCCESS,
    borderTopLeftRadius: 20,
    borderBottomLeftRadius: 20,
  },
  bankTransfer: {
    display: 'flex',
    flexDirection: 'row',
    gap: 5,
    alignItems: 'center'
  },
  amountAndDate: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center'
  },
});

export default styles;
