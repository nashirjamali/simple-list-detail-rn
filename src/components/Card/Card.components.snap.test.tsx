import React from 'react';
import {render} from '@testing-library/react-native';

import {CARD_VARIANTS} from '../../constants';
import type {CardProps} from '../../@types';

import Card from './Card.components';

describe('Card Component', () => {
  const defaultProps: CardProps = {
    amount: 1000000,
    beneficiaryBank: 'BCA',
    beneficiaryName: 'John Doe',
    completedAt: '2023-12-25T00:00:00Z',
    senderBank: 'Mandiri',
    status: CARD_VARIANTS.SUCCESS,
    onPress: jest.fn(),
  };

  it('renders correctly with success status', () => {
    const {toJSON} = render(<Card {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with pending status', () => {
    const {toJSON} = render(
      <Card {...defaultProps} status={CARD_VARIANTS.PENDING} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
