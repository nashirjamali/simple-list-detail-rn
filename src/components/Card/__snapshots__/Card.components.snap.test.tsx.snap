// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component renders correctly with pending status 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 20,
      "display": "flex",
      "flexDirection": "row",
      "opacity": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#FB6544",
        "borderBottomLeftRadius": 20,
        "borderTopLeftRadius": 20,
        "width": 10,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 20,
        "width": "100%",
      }
    }
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "gap": 5,
          }
        }
      >
        <Text
          style={
            {
              "color": "#121212",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          MANDIRI
        </Text>
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={20}
          bbWidth={20}
          fill="none"
          focusable={false}
          height={20}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 20,
                "width": 20,
              },
            ]
          }
          vbHeight={24}
          vbWidth={24}
          width={20}
        >
          <RNSVGGroup
            fill={null}
            propList={
              [
                "fill",
              ]
            }
          >
            <RNSVGPath
              d="M13.233 16.457a.75.75 0 101.034 1.086l-1.034-1.086zm6.284-3.914a.75.75 0 10-1.034-1.086l1.034 1.086zm-1.034 0a.75.75 0 101.034-1.086l-1.034 1.086zm-4.216-6.086a.75.75 0 10-1.034 1.086l1.034-1.086zM19 12.75a.75.75 0 000-1.5v1.5zm-14-1.5a.75.75 0 000 1.5v-1.5zm9.267 6.293l5.25-5-1.034-1.086-5.25 5 1.034 1.086zm5.25-6.086l-5.25-5-1.034 1.086 5.25 5 1.034-1.086zM19 11.25H5v1.5h14v-1.5z"
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "fill",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
        <Text
          style={
            {
              "color": "#121212",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          BCA
        </Text>
      </View>
      <Text
        style={
          {
            "color": "#121212",
            "fontSize": 14,
          }
        }
      >
        JOHN DOE
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            {
              "color": "#121212",
              "fontSize": 14,
            }
          }
        >
          Rp1.000.000
        </Text>
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={20}
          bbWidth={20}
          fill="none"
          focusable={false}
          height={20}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 20,
                "width": 20,
              },
            ]
          }
          vbHeight={24}
          vbWidth={24}
          width={20}
        >
          <RNSVGGroup
            fill={null}
            propList={
              [
                "fill",
              ]
            }
          >
            <RNSVGPath
              d="M12 9.5a2.5 2.5 0 110 5 2.5 2.5 0 010-5z"
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "fill",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
        <Text
          style={
            {
              "color": "#121212",
              "fontSize": 14,
            }
          }
        >
          25 Desember 2023
        </Text>
      </View>
    </View>
    <View>
      <View
        style={
          {
            "borderColor": "#FB6544",
            "borderRadius": 10,
            "borderWidth": 2,
            "justifyContent": "center",
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#121212",
              "fontSize": 14,
              "fontWeight": "bold",
              "textAlign": "center",
            }
          }
        >
          Pengecekan
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Card Component renders correctly with success status 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 20,
      "display": "flex",
      "flexDirection": "row",
      "opacity": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#55B585",
        "borderBottomLeftRadius": 20,
        "borderTopLeftRadius": 20,
        "width": 10,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 20,
        "width": "100%",
      }
    }
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "gap": 5,
          }
        }
      >
        <Text
          style={
            {
              "color": "#121212",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          MANDIRI
        </Text>
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={20}
          bbWidth={20}
          fill="none"
          focusable={false}
          height={20}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 20,
                "width": 20,
              },
            ]
          }
          vbHeight={24}
          vbWidth={24}
          width={20}
        >
          <RNSVGGroup
            fill={null}
            propList={
              [
                "fill",
              ]
            }
          >
            <RNSVGPath
              d="M13.233 16.457a.75.75 0 101.034 1.086l-1.034-1.086zm6.284-3.914a.75.75 0 10-1.034-1.086l1.034 1.086zm-1.034 0a.75.75 0 101.034-1.086l-1.034 1.086zm-4.216-6.086a.75.75 0 10-1.034 1.086l1.034-1.086zM19 12.75a.75.75 0 000-1.5v1.5zm-14-1.5a.75.75 0 000 1.5v-1.5zm9.267 6.293l5.25-5-1.034-1.086-5.25 5 1.034 1.086zm5.25-6.086l-5.25-5-1.034 1.086 5.25 5 1.034-1.086zM19 11.25H5v1.5h14v-1.5z"
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "fill",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
        <Text
          style={
            {
              "color": "#121212",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          BCA
        </Text>
      </View>
      <Text
        style={
          {
            "color": "#121212",
            "fontSize": 14,
          }
        }
      >
        JOHN DOE
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            {
              "color": "#121212",
              "fontSize": 14,
            }
          }
        >
          Rp1.000.000
        </Text>
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={20}
          bbWidth={20}
          fill="none"
          focusable={false}
          height={20}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 20,
                "width": 20,
              },
            ]
          }
          vbHeight={24}
          vbWidth={24}
          width={20}
        >
          <RNSVGGroup
            fill={null}
            propList={
              [
                "fill",
              ]
            }
          >
            <RNSVGPath
              d="M12 9.5a2.5 2.5 0 110 5 2.5 2.5 0 010-5z"
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "fill",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
        <Text
          style={
            {
              "color": "#121212",
              "fontSize": 14,
            }
          }
        >
          25 Desember 2023
        </Text>
      </View>
    </View>
    <View>
      <View
        style={
          {
            "backgroundColor": "#55B585",
            "borderRadius": 10,
            "justifyContent": "center",
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 14,
              "fontWeight": "bold",
              "textAlign": "center",
            }
          }
        >
          Berhasil
        </Text>
      </View>
    </View>
  </View>
</View>
`;
